beacon-chain/rpc/prysm/v1alpha1/validator/proposer_attestations.go

```diff
@@ -209,6 +219,142 @@ func (a proposerAtts) sortByProfitabilityUsingMaxCover() (proposerAtts, error) {
 	return sortedAtts, nil
 }
 
+// sort attestations as follows:
+//
+//   - all attestations selected by max-cover are ordered before leftover attestations
+//   - within selected/leftover, attestations are ordered by slot, with higher slot coming first
+//   - within a slot, all top attestations (one per committee) are ordered before any second-best attestations, second-best before third-best etc.
+//   - within top/second-best/etc. attestations (one per committee), attestations are ordered by bit count, with higher bit count coming first
+func (a proposerAtts) sort() (proposerAtts, error) {
+	if len(a) < 2 {
+		return a, nil
+	}
+
+	type slotAtts struct {
+		candidates map[primitives.CommitteeIndex]proposerAtts
+		selected   map[primitives.CommitteeIndex]proposerAtts
+		leftover   map[primitives.CommitteeIndex]proposerAtts
+	}
+
+	// Separate attestations by slot, as slot number takes higher precedence when sorting.
+	// Also separate by committee index because maxcover will prefer attestations for the same
+	// committee with disjoint bits over attestations for different committees with overlapping
+	// bits, even though same bits for different committees are separate votes.
+	var slots []primitives.Slot
+	attsBySlot := map[primitives.Slot]*slotAtts{}
+	for _, att := range a {
+		slot := att.GetData().Slot
+		ci := att.GetData().CommitteeIndex
+		if _, ok := attsBySlot[slot]; !ok {
+			attsBySlot[slot] = &slotAtts{}
+			attsBySlot[slot].candidates = make(map[primitives.CommitteeIndex]proposerAtts)
+			slots = append(slots, slot)
+		}
+		attsBySlot[slot].candidates[ci] = append(attsBySlot[slot].candidates[ci], att)
+	}
+
+	var err error
+	for _, sa := range attsBySlot {
+		sa.selected = make(map[primitives.CommitteeIndex]proposerAtts)
+		sa.leftover = make(map[primitives.CommitteeIndex]proposerAtts)
+		for ci, committeeAtts := range sa.candidates {
+			sa.selected[ci], sa.leftover[ci], err = committeeAtts.sortByProfitabilityUsingMaxCover_committeeAwarePacking()
```

Comment:
Is there a test where we check for leftovers? I think that with the parameters passed to `MaxCover` the leftover map will always be empty unless there are contained aggregation bits, in which case anyway we would drop them. I suspect that with these parameters you can simply ignore the leftovers from MaxCover and just consider the selected ones.

Reply:
```
atts := getAtts([]testData{
			{4, bitfield.Bitlist{0b00000001, 0b1}},
			{4, bitfield.Bitlist{0b11100001, 0b1}},
			{1, bitfield.Bitlist{0b11000000, 0b1}},
			{2, bitfield.Bitlist{0b11100000, 0b1}},
			{4, bitfield.Bitlist{0b10000011, 0b1}},
			{4, bitfield.Bitlist{0b11111000, 0b1}},
			{1, bitfield.Bitlist{0b11100000, 0b1}},
			{3, bitfield.Bitlist{0b11000000, 0b1}},
		})
		want := getAtts([]testData{
			{4, bitfield.Bitlist{0b11111000, 0b1}},
			{4, bitfield.Bitlist{0b10000011, 0b1}},
			{3, bitfield.Bitlist{0b11000000, 0b1}},
			{2, bitfield.Bitlist{0b11100000, 0b1}},
			{1, bitfield.Bitlist{0b11100000, 0b1}},
			{4, bitfield.Bitlist{0b11100001, 0b1}},
			{4, bitfield.Bitlist{0b00000001, 0b1}},
			{1, bitfield.Bitlist{0b11000000, 0b1}},
		})
```
This is an example of a test where we check that left over aggregates are sorted last. I agree that in all such tests each leftover atts has bits that are already covered.